include/roboteam_ai/utilities/IOManager.h

```diff
@@ -32,54 +28,29 @@ using namespace rtt::world;
 
 class IOManager {
    private:
-    proto::World worldMsg;
-    proto::SSL_GeometryData geometryMsg;
-    proto::RobotFeedback robotFeedbackMsg;
-    proto::SSL_Referee refDataMsg;
-    proto::DemoRobot demoInfoMsg;
-
-    std::unordered_map<uint8_t, proto::RobotFeedback> feedbackMap;
-
-    proto::Subscriber<proto::World> *worldSubscriber;
-    void handleWorldState(proto::World &world);
-
-    proto::Subscriber<proto::SSL_GeometryData> *geometrySubscriber;
-    void handleGeometry(proto::SSL_GeometryData &geometryData);
+    proto::State state;
+    proto::Subscriber<proto::State> *worldSubscriber;
+    void handleState(proto::State &state);
 
-    proto::Subscriber<proto::SSL_Referee> *refSubscriber;
-    void handleReferee(proto::SSL_Referee &refData);
-
-    proto::Subscriber<proto::RobotFeedback> *feedbackSubscriber;
-    void handleFeedback(proto::RobotFeedback &feedback);
-
-    proto::Publisher<proto::RobotCommand> *robotCommandPublisher;
+    proto::Publisher<proto::AICommand> *robotCommandPublisher;
     proto::Publisher<proto::Setting> *settingsPublisher;
 
     rtt::ai::Pause *pause;
 
+    rtt::networking::PairReceiver<16970> *  central_server_connection;
    public:
     ~IOManager();
     explicit IOManager() = default;
-    void publishRobotCommand(proto::RobotCommand cmd, rtt::world::World const* world);
+    void publishAllRobotCommands(const std::vector<proto::RobotCommand>& vector);
     void publishSettings(proto::Setting setting);
+    void handleCentralServerConnection();
     void init(int teamId);
-    proto::World getWorldState();
-    proto::SSL_GeometryData getGeometryData();
-    std::unordered_map<uint8_t, proto::RobotFeedback> getFeedbackDataMap();
-
-    proto::SSL_Referee getRefereeData();
-    const proto::DemoRobot &getDemoInfo();
-
-    static std::mutex worldStateMutex;
-    static std::mutex geometryMutex;
-    static std::mutex robotFeedbackMutex;
-    static std::mutex refereeMutex;
-    static std::mutex demoMutex;
+    proto::State getState();
 
-    bool hasReceivedGeom = false;
+    mutable std::shared_mutex stateMutex;
```

Comment:
why is this a shared_mutex and please don't mark it mutable, keep the class as mutable instead.

Reply:
you shouldn't even need this to be `mutable`, if you mark the class as `const` you can't modify the data anyway and then you don't need a mutex since it's read-only.

Reply:
Ah nevermind i see what's going on now, although I can't say I like it :P

Reply:
No, you're right this is not necessary. I changed it during testing because I thought there was a deadlock, but it turned out to be a ZMQ blocking-call related problem.

Reply:
I changed it back to a normal std::mutex again.

Reply:
I'll change it back to a normal std::mutex again.

Reply:
Sure.